### ISSUE TITLE
Incorrect closure capture analysis

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#20155]
```rust
        let b = Some(Box::new(0));
        || {
            if let Some(b) = b {
                let _move = b;
            }
        };
```
we incorrectly report an error on the move of `b` here. We capture `b` by reference, then complain we can't move it. We should be capturing it by value.

**Comment by ShoyuVanilla** [comment]
I thought I had fixed it with #20039 but it seems there's more 😅

**Comment by Veykril** [comment]
I think this even got surfaced only by that PR, because I noticed thise in the r-a repo and I am pretty sure it didn't use to error there in the past :D

**Comment by Veykril** [comment]
Also triggers here for example https://github.com/rust-lang/rust-analyzer/blob/a3dc41bf9753b1595742aed47321561d1a1e93f8/crates/hir-def/src/expr_store/lower.rs#L1400

**Comment by ShoyuVanilla** [comment]
I found the cause (we are not visiting bind patterns in let expr/stmts, on contrary to rustc) and will fix when I'm back from work 😄
